guides-markdown/cltv.md

```diff
@@ -0,0 +1,624 @@
+must match a public key or public key hash that is on the execution stack.
+
+An output that is locked with CLTV works more or less the same way but adds another
+condition that before the signature is even accepted, a certain amount of time must have passed.
+
+In pseudo-code, the script will look like this:
+
+```
+Locktime (in blocks or Unix epoch time)
+Check if locktime is less than nLocktime of transaction; execution fails if not
+Check public key hash matches
+Check if the signature validates
+```
+
+In Bitcoin, this locking script will look something like this:
+```
+<some amt of time> CHECKLOCKTIMEVERIFY DROP DUP HASH160
+<Redeeming Public Key Hash> EQUALVERIFY CHECKSIG
+```
+
+To learn more about CLTV and how it works in Bitcoin, checkout [chapter 07 in
+Mastering Bitcoin](https://github.com/bitcoinbook/bitcoinbook/blob/develop/ch07.asciidoc). See [Chapter 06](https://github.com/bitcoinbook/bitcoinbook/blob/develop/ch06.asciidoc#transaction-scripts-and-script-language)
+for more information on how the stack and scripts operate.
+
+This guide will be broken into two parts. First we will walk through
+creating and redeeming a mock transaction with no real coins. Next we will do the
+same thing on a live regtest network using the bcoin wallet and API.
+
+(If you want to see the full code, checkout the repo [here](https://github.com/bucko13/cltv))
+
+## The Guide - Mock Transaction
+Before moving on to the transaction there is some setup that will be useful
+for both approaches. Note that this guide assumes use with Segwit transactions so
+inputs and redeem scripts are stored in the witness field.
+
+### Setup
+Let's first setup the constants and import the modules we will need to create
+and spend our transactions.
+
+```javascript
+// first the modules we will be using from bcoin
+const {
+  Amount,
+  Coin,
+  KeyRing,
+  MTX,
+  Network,
+  Outpoint,
+  Script,
+  ScriptNum,
+  Stack
+} = require('bcoin');
+
+// some NodeJS helpers
+const fs = require('fs');
+const assert = require('assert');
+
+// a helper object with information
+// about the regtest network
+const network = Network.get('regtest');
+```
+
+### Creating the Script
+Let's create a function that takes a locktime and a publicKeyHash
+that corresponds to the private key that can redeem the transaction
+and returns an object of type `Script`. This script is what will
+ultimately be used to lock our UTXO, i.e. the locking script.
+
+Note that for CLTV, the locktime can either be a block height _or_
+Unix Epoch timestamp (seconds since Jan-1-1970) if above 500 million (
+[See BIP-065 Specs](https://github.com/bitcoin/bips/blob/master/bip-0065.mediawiki)).
+
+```javascript
+/**
+ * @param {String} locktime - Time that the script can not
+ * be redeemed before
+ * @param {Buffer} public key hash
+ * @returns {Script}
+**/
+function createScript(locktime='100', publicKeyHash) {
```

Comment:
Is it possible to use the default value of `locktime` in this case when its the first argument?

```
createScript(,'xxx')
```

I don't think that you should give it a default value and that you should `assert` that it was passed in

Reply:
Good call!